RFC: Add Store Pool Service [ci skip]

BramGruneir opened the pull request:
RFC to add a new store monitoring service.

Works in conjunction with the other 2 open RFCs #2153 and #2171.

BramGruneir committed:
RFC: Add Store Pool Service [ci skip]

BramGruneir commented:
I'd be down for a better name than `Monitor`. It seems very overused.

petermattis commented on docs/RFCS/node_monitor.md:
```diff
@@ -0,0 +1,69 @@
+   - Gossip requires a good refactoring and I can see this as one of the first
+   steps to do so. This will mean that you can get node lists from somewhere
+   other than gossip and it will include more details as well. Adding these
+   calculations into gossip seems cumbersome.
```
Why is adding to gossip cumbersome? It also seems cumbersome for clients of gossip and the monitor service to have to query two different locations to get the full picture about a node (health + stats). The monitor service can certainly copy this data from gossip, but then we'd have the data duplicated.

BramGruneir replied:
The data itself is very small so duplication won't be an issue. There's no need to query two different locations, this will maintain a list of all nodes stats and health.

There seems to be a general consensus around the office that gossip needs to be refactored and tailored down to its bare essentials (see #2170). Adding in logic about the node health and possibly which node a new replica should be added to (or removed from) seems like a variation on what is mostly a system designed for communication.

petermattis replied:
Ok.

petermattis commented on docs/RFCS/node_monitor.md:
```diff
@@ -0,0 +1,69 @@
+
+# Summary
+
+Add a new `Monitor` service on each node that monitors all the other nodes and
```
`NodeStatus` service?

BramGruneir replied:
I just didn't want to use the term status again, since that's very involved with the status server.  But monitor is also used all over too.
Perhaps going with `NodePool` might be better.  Which would make sense if we add some of the decision making logic to it.

BramGruneir commented:
Changed this a bit.
- After a discussion with @mrtracy, nodes are not the correct scope for rebalancing, stores are. And it's possible for a single store to have an issue while the rest of them on a node are working correctly.
- Added a comment about moving some (or perhaps all) of the allocator functions into it.
, it's now called `StorePool`, but I'm not sold on that name either.  I just want to avoid anything with status or monitor in the title.

petermattis commented:
Can you expand on why store is the right level to do allocation at? One advantage to allocating at the node level is that the node itself can then make the local decision of where to place the range using more up to date information than is gossiped.

BramGruneir commented:
Of course. 
- Right now we only gossip capacity for stores, not nodes. 
- If you look at the allocator, it functions on the level of the store as well.
- The store descriptor includes capacity information while the node descriptor does not.  So it would have to either be added or inferred.
- Most setups will have only one store per node, so the extra logic might go unused most of the time.
- Finally, there's the possibility that an hdd/ssd is acting up or slowed causing a single store in a node to malfunction.  In that case, we'll be able to isolate that store but keep the node's other stores active.

However, we could code around all of these issues easily enough if there was a compelling reason to do so. The only one I can think of is the one you've mentioned about giving the node more control over the ranges on its stores. And we don't even rebalance ranges between stores in a node right now (and I don't think there's a plan do so).

petermattis commented:
I think allocating at the store level is correct for now I just wanted you to point out that allocating at that level is not obviously the correct long term choice.

BramGruneir commented:
Gotcha, I'll add a note to the RFC about just that.

BramGruneir commented:
Added.

vivekmenezes commented on docs/RFCS/store_pool.md:
```diff
@@ -0,0 +1,79 @@
+- Feature Name: Store Pool
```
Store health?

BramGruneir replied:
If it only was a repository for the health of all stores, then sure, but once we add any other logic, it doesn't make sense.

vivekmenezes replied:
Yeah, we know how that happens :-)

vivekmenezes commented on docs/RFCS/store_pool.md:
```diff
@@ -0,0 +1,79 @@
+if that store has not been updated.
+If the store hasn't been updated, it will mark it as dead.
+Then it will calculate the next `timeUntilNextDead` to wake up the service.
+
```
This StoreHealth service will also be responsible for marking a store as healthy right?

BramGruneir replied:
Indeed. As part of the update from the gossip trigger.  I'll add something about that.

spencerkimball commented on docs/RFCS/store_pool.md:
```diff
@@ -0,0 +1,79 @@
+gossip.
+
+`StorePool` will maintain a map of store IDs to store descriptors and a variety
+of heath statistic about the store. It will also maintain a `lastUpdatedTime`
```
s/heath/health/

BramGruneir replied:
fixed

spencerkimball commented on docs/RFCS/store_pool.md:
```diff
@@ -0,0 +1,79 @@
+of heath statistic about the store. It will also maintain a `lastUpdatedTime`
+which will be set whenever a store descriptor is updated. To maintain this map,
+a callback from gossip for store descriptors will be added. When this
+`lastUpdatedTime` is longer than the `timeoutUnntilStoreDead`, the store is
```
s/timeoutUnntil/timeoutUntil/

BramGruneir replied:
fixed

spencerkimball commented on docs/RFCS/store_pool.md:
```diff
@@ -0,0 +1,79 @@
+a callback from gossip for store descriptors will be added. When this
+`lastUpdatedTime` is longer than the `timeoutUnntilStoreDead`, the store is
+considered dead and any replicas on this store may be removed. Note that this
+removal will be done elsewhere.
```
s/that this removal will be/that the work to remove replicas is/

BramGruneir replied:
fixed.

spencerkimball commented on docs/RFCS/store_pool.md:
```diff
@@ -0,0 +1,79 @@
+
+# Alternatives
+
+1. Instead of creating a new store monitoring service, add all of this into
```
This one's a non-starter. I'd just eliminate it.

BramGruneir replied:
Actually, this was discussed more than once. And was briefly the direction I was going with. It is indeed the wrong choice, but I'd like to leave it in as an alternative.

spencerkimball commented:
LGTM

bdarnell commented on docs/RFCS/store_pool.md:
```diff
@@ -0,0 +1,89 @@
+gossip.
+
+`StorePool` will maintain a map of store IDs to store descriptors and a variety
+<<<<<<< Updated upstream
```
There's some merge cruft here.

BramGruneir replied:
Fixed.

bdarnell commented:
LGTM

BramGruneir commented:
Thanks everyone for taking a look.
